### DISCUSSION TITLE
Formalize PR state storage

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#27385] [Suggest an Idea]
### Describe the proposed change(s).

There are some cases where it would be useful for Renovate to store "additional" state in PRs. For example:
- For immortal PRs, the exact list of updates contained within, so that the PR can be "mortal" for as long as it is unique
- Storing the original list of labels/reviewers/assignees/etc, e.g. #25340
- Recording which version of Renovate was used for the latest commit, PR creation, PR update, etc.
- Allowing third party systems to parse Renovate PRs and _know_ what's in it (not just _guess_)

For platforms which support hidden HTML comments, we now embed some of the above information into the PR body: https://github.com/renovatebot/renovate/blob/96d174b460707ba07ff81d50887ebfe1848a7f1f/lib/modules/platform/types.ts#L54-L58

But here I want to discuss:
- Is this the best way?
- What about platforms where it can't be hidden - is it still worth it to show this "debug" data even if it's a bit ugly?
- Is there some such data which is _useful_ to end users to see and should be intentionally non-hidden? e.g. we already show the list of updates in markdown tables because that's important, although for example the lists of labels/assignees/reviewers would not be useful
- Are there platforms with too little PR body characters that it starts impacting how much "real" body we can use?
- Is it easier to embed the data in git commits instead of the PR itself? e.g. using [git notes](https://tylercipriani.com/blog/2022/11/19/git-notes-gits-coolest-most-unloved-feature/). It seems great for us but would be less useful for third parties as no platform treats git notes as a first class concept (e.g. accessible via API)

**Comment by rarkins** [comment]
## PR body character limitations and HTML hidden comment support

| Platform         | Character limit | HTML | Hidden comments |
|------------------|-----------------|------|-----------------|
| Azure DevOps     | 4,000           | ✅    | ❌               |
| Bitbucket Cloud  | 50,000          | ❌    | ❌               |
| Bitbucket Server | 30,000          | ❌    | ❌               |
| Gerrit           | 16,384          | ❌    | ❌               |
| Gitea            | 1,000,000       | ✅    | ✅               |
| GitHub           | 60,000          | ✅    | ✅               |
| GitLab           | 1,000,000       | ✅    | ✅               |

**Comment by rarkins** [comment]
## Debug data not interesting to users:

These are useful to Renovate internally in order to know if they _changed_ in config since PR creation. For example if the target (base) branch has changed, we want to update the PR to match. But if a _user_ changed the target branch then we want to consider it to be _modified_ and not revert it back.

- Assignees
- Base branch
- Labels
- Reviewers

We haven't found a great use for these yet but it seems useful to record the version of Renovate used:

- Versions of Renovate

## Data also useful to users:

- List of updates

Today this is displayed in a pretty manner in markdown tables. It could also be displayed in user or machine-readable formats in git commit bodies. Either way, it's ideal that Renovate can parse it reliably. Are human readable tables good enough for that?

**Comment by HonkingGoose** [comment]
## Show Renovate version in plain text

> Recording which version of Renovate was used for the latest commit, PR creation, PR update, etc.

We already show some parts of this, don't we? But hidden behind base64 encoding. I'd say show this in plaintext, this makes it easier for those reporting a bug with Renovate to say what version they're using.

## Show debug data, even if ugly

> What about platforms where it can't be hidden - is it still worth it to show this "debug" data even if it's a bit ugly?

If we need to choose between functional and nice looking, I'll take functional. 😄 

## Show timing statistics and cache hit or miss info?

> Is there some such data which is useful to end users to see and should be intentionally non-hidden? e.g. we already show the list of updates in markdown tables because that's important, although for example the lists of labels/assignees/reviewers would not be useful

How about putting some timing statistics in the PR body? For example:

- Start of job: `00:00:00`
- End of job: `00:05:00`
- Total time in minutes: `5 minutes 0 seconds`
- Maybe show cache hit / cache miss information?

I agree that the lists of labels and assignees and reviewers is not needed, because the Git hosting platform will show that information, or show it via a API that bots can read.

This new issue seems related:

- #27035

## Decide on information priority for low character limit platforms

> Are there platforms with too little PR body characters that it starts impacting how much "real" body we can use?

You'll probably need special handling for the Azure and Gerrit platforms, as they only allow 4000 or 16.384 characters according to your table.

We need to figure out and decide what priority to give each field, so we're showing the most important stuff first, and drop less important stuff, if we're running out of characters.

## If we're going to use Git notes write good docs

> Is it easier to embed the data in git commits instead of the PR itself? e.g. using [git notes](https://tylercipriani.com/blog/2022/11/19/git-notes-gits-coolest-most-unloved-feature/). It seems great for us but would be less useful for third parties as no platform treats git notes as a first class concept (e.g. accessible via API)

I've never heard of git notes... 🙃 So if you decide to do this, this needs to be properly explained in the docs. 😄 

## Current Markdown table not accessible

> - List of updates
> 
> Today this is displayed in a pretty manner in markdown tables. It could also be displayed in user or machine-readable formats in git commit bodies. Either way, it's ideal that Renovate can parse it reliably. Are human readable tables good enough for that?

I like the table-view of the updates, but I can _see_ them well. I think our current Markdown-only tables are _not accessible for blind users_. [^1] [^2] It's crucial that screen readers know something is a _data table_:

> A critical step toward creating an accessible data table is to designate row and/or column headers.

The WebAIM data table examples use _HTML tables_, with `<table>`, `<td>`, `<td>` and `<caption>` elements. The Markdown version lacks all those descriptive elements.

It looks like we would need to start writing a proper HTML table, so that everybody can use Renovate's PR body tables.

[^1]: [WebAIM, creating accessible tables, page 1](https://webaim.org/techniques/tables/)
[^2]: [WebAIM, creating accessible tables, page 2](https://webaim.org/techniques/tables/data)
